File: solutions/ocaml/035-permutations.ml
(*
  Problem 35: Permutations (LeetCode 46)
  Difficulty: Med
  Language: OCaml
*)
let rec permute = function
  | [] -> [[]]
  | xs ->
    List.concat_map (fun x ->
      let rest = List.filter (fun y -> y <> x) xs in
      List.map (fun p -> x :: p) (permute rest)
    ) xs

File: solutions/haskell/035-permutations.hs
{-
  Problem 35: Permutations (LeetCode 46)
  Difficulty: Med
  Language: Haskell
-}
import Data.List (delete)

permute :: Eq a => [a] -> [[a]]
permute [] = [[]]
permute xs = concatMap (\x -> map (x:) (permute (delete x xs))) xs
